### ISSUE TITLE
let_and_return false positive

### ISSUE THREAD

**Issue by JohnTheCoolingFan** [rust-lang/rust-clippy#9794]
### Summary

The lint says to remove binding to a variable and immediately returning the variable, yet when I do that the code fails to compile and suggests the original code that clippy marks with a warning.

### Lint Name

let_and_return

### Reproducer

I tried this code:

```rust
let channel = {
    let modlist_ic_data = guild_config_read.info_channels_data(InfoChannelType::ModList);
    let channel = modlist_ic_data.read().await.channel_id; channel
};

```

I saw this happen:

```
warning: returning the result of a `let` binding from a block
   --> src/command_groups/factorio.rs:113:64
    |
113 |         let channel = modlist_ic_data.read().await.channel_id; channel
    |         ------------------------------------------------------ ^^^^^^^
    |         |
    |         unnecessary `let` binding
    |
    = note: `#[warn(clippy::let_and_return)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_and_return
help: return the expression directly
    |
113 -         let channel = modlist_ic_data.read().await.channel_id; channel
113 +          modlist_ic_data.read().await.channel_id
    |

warning: `rusted-motorchik` (bin "rusted-motorchik") generated 1 warning
```

I expected to see this happen:
No lint at all since doing what clippy suggested fails to compile

### Version

```text
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0
```

### Additional Labels

@rustbot label D-invalid-suggestion

**Comment by Alexendoo** [comment]
Could you show the error you get with the suggestion applied? And if the code is publicly available a link to it would be great

**Comment by JohnTheCoolingFan** [comment]
Here's the error I get when applying the suggestion:
```
error[E0597]: `modlist_ic_data` does not live long enough
   --> src/command_groups/factorio.rs:113:9
    |
113 |         modlist_ic_data.read().await.channel_id
    |         ^^^^^^^^^^^^^^^^^^^^^^------
    |         |
    |         borrowed value does not live long enough
    |         a temporary with access to the borrow is created here ...
114 |     };
    |     -- ... and the borrow might be used here, when that temporary is dropped and runs the `Drop` code for type `tokio::sync::RwLockReadGuard`
    |     |
    |     `modlist_ic_data` dropped here while still borrowed
    |
    = note: the temporary is part of an expression at the end of a block;
            consider forcing this temporary to be dropped sooner, before the block's local variables are dropped
help: for example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block
    |
113 |         let x = modlist_ic_data.read().await.channel_id; x
    |         +++++++                                        +++

For more information about this error, try `rustc --explain E0597`.
error: could not compile `rusted-motorchik` due to previous error
```
The code is available on github: https://github.com/JohnTheCoolingFan/rusted-motorchik/blob/master/src/command_groups/factorio.rs#L111-L114

**Comment by ada4a** [comment]
This no longer lints on edition 2021 as per https://github.com/rust-lang/rust-clippy/issues/13962#issuecomment-3148713194:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=eb21932e9364434018008b30140b59e9

@rustbot label S-fixed
